File: prc/tc/g.tc.py
#!/usr/bin/env python

"""%s

purpose:

usages:

  %s  VVV (verifile) -s SSS.YYYY (stmid) -m MMM (Model)

  -P prodopt
  -I interactive mode
  
examples:
%s ops.2004.shem -s 04S.2004 -m ukm
"""

import os
import sys
import posix
import posixpath
import getopt

import string
import glob

from time import sleep
import TCw2 as TC


mflibdir=TC.MfLibrary

sys.path.append(mflibdir)
import mf


convertexe='/pcmdi/tenki_opt/linux/bin/convert'


#
#  defaults
#

stmopt=None
veriname=None
modelopt=None
taumax=120
hhopt='00_12'
hhopt='all'
interact=None

popt='ops'
batchopt=None
prodopt=None

ropt=''

verb=0

curdtg=mf.dtg()
curtime=mf.dtg('curtime')
curdir=posix.getcwd()
pyfile=sys.argv[0]

narg=len(sys.argv)-1

#
# veriname is always arg #1
#

#
# options using getopt
#

if(narg > 0):

    veriname=sys.argv[1]

    try:
        (opts, args) = getopt.getopt(sys.argv[2:], "s:m:p:v:BNPIV")

    except getopt.GetoptError:
        mf.usage(__doc__,pyfile,curdtg,curtime)
        print "EEE invalid getopt opt"
        sys.exit(2)

    for o, a in opts:
        if o in ("-s","--vopt"): stmopt=a
        if o in ("-m","--vopt"): modelopt=a
        if o in ("-p","--vopt"): popt=a
        if o in ("-B","--vopt"): batchopt=1
        if o in ("-P","--vopt"): prodopt=1
        if o in ("-N","--vopt"): ropt='norun'
        if o in ("-I","--vopt"): interact=1
        if o in ("-V",""): verb=1

else:
    mf.usage(__doc__,pyfile,curdtg,curtime)
    sys.exit(1)

if(not(veriname) or not(stmopt) or not(modelopt)):
    print "EEE run using -v veriname -s stmopt -m model"
    sys.exit()

try:
    (tstmid,tyear)=stmopt.split('.')
except:
    print "EEE unable to split stmopt use -s SSS.YYYY"
    sys.exit()

tmodel=modelopt

tt=veriname.split('.')
if( (len(tt) == 3) and (tt[0] == 'ops') ):
    opsyear=tt[1]
    opshemi=tt[2]
else:
    opsyear=None
    opshemi=None

tcnames=TC.GetTCnamesHash(tyear)
tcstats=TC.GetTCstatsHash(tyear)

btdir=TC.BtDir
mddir=TC.MdeckDir
ftdir=TC.FcOpsDir
vddir=TC.VdeckDir

veridir=TC.VeriDir
pltdir=TC.PltOpsTrackDir

if(opsyear and opshemi):
    pltdir="%s/%s.%s"%(pltdir,opsyear,opshemi)
    print 'pp ',pltdir
    if(not(os.path.isdir(pltdir))):
        try:
            os.mkdir(pltdir)
            print 'DDDDDD made: ',pltdir
        except:
            print "EEE unable to mkdir: "%(pltdir)
            sys.exit()
    
bid=tstmid[2:3]
bname=TC.Basin1toBasin3[bid]
mdmask="%s/%s/MD.%s.%s.%s.*.txt"%(mddir,tyear,bname,tyear,tstmid)
mdpaths=glob.glob(mdmask)
print mdmask
if(len(mdpaths) == 0):
    print 'EEE no bts for mdmask: ',mdmask
    sys.exit()
else:
    mdpath=mdpaths[0]
    
vimask="%s/%s/vdeck.%s.%s.%s.%s.*txt"%(vddir,tyear,bname,tyear,tstmid,tmodel)
vipaths=glob.glob(vimask)
if(len(vipaths) == 0):
    print 'EEE no vdeck for vimask: ',vimask
    sys.exit()
else:
    vipath=vipaths[0]

ftgpath="/tmp/g.w2.tc.fc.txt"
pystat="TCfcStat"
verirule='jtwc.hetero'
#verirule='jtwc.mod.hetero'

#
#  synoptic time and last veri tau
#

synopt='syn00_12'

if(bid == 'P' or bid == 'S'):
    opt3='48veri'
    taufcveristat='48'
else:
    opt3='72veri'
    taufcveristat='72'


stipath="%s/tc.veri.sum.%s.%s.py"%(veridir,veriname,verirule)
vpath="%s/tc.veri.%s.tc.ctl"%(vddir,tyear);


if(not(os.path.exists(stipath))):
    print "EEE  stipath: %s does not exist"%(stipath)
    sys.exit()
else:
    cmd="cp %s %s.py"%(stipath,pystat)
    mf.runcmd(cmd)
    from TCfcStat import tcfc


try:
    tstmname=tcnames[tyear,tstmid]
except:
    tstmname='NoName'

try:
    tstmstat=tcstats[tyear,tstmid]
    tstmtype=tstmstat[0]
except:
    tstmtype='TC'
    print "EEEEEEEEEEEEEEEE problem with tcstats: %s %s"%(tyear,tstmid)
    for tcstat in tcstats:
        print 'qqq ',tcstat
    print tcstats[tyear,tstmid]
    sys.exit()

try:
    tcfcstat=tcfc[tyear,tstmid]
except:
    tcfcstat=None

tstmtype=tstmtype.strip()

TC.TcFcStatLegend(tcfcstat,tmodel,bid,taufcveristat,verb=0)
    

if(verb):
    print "TTT     tstmid:  %s"%(tstmid)
    print "TTT   tstmtype:  %s"%(tstmtype)
    print "TTT   tcfcstat: ",tcfcstat
    
    print "DDD      mddir:  %s"%(mddir)
    print "DDD      ftdir:  %s"%(ftdir)
    print "DDD      vddir:  %s"%(vddir)

    print "PPP     mdpath:  %s"%(mdpath)
    print "PPP     vipath:  %s"%(vipath)
    
#
# grep out to speed up get model fc tracks
#

cards=open(vipath).readlines()

#for card in cards:
#    print card
#sys.exit()
#cmd="grep -h -s %s %s | grep %s > %s"%(stmopt,vipath,tmodel,ftipath)
#mf.runcmd(cmd,ropt)
#cards=open(ftipath).readlines()

tcs={}
dtgs=[]

for c in cards:
    (model,tau,sid,snum,sbasin,bdtg,vdtg,flat,flon,blat,blon,
     fe,bvmax,fvmax,cte,ate,fve,fveu)=TC.vcparse(c)
    stmid=sid.split('.')[0]
    stmyyyy=sid.split('.')[1]
    if(model == tmodel and stmid == tstmid and stmyyyy == tyear):
        #print 'qqqqqqqqqqqqqqqq ', model,tmodel,tau,bdtg,sid,stmid,tstmid,stmyyyy,flat,flon
        dtgs.append(bdtg)
        try:
            tcs[bdtg].append([model,tau,flat,flon,fvmax,fe])
        except:
            tcs[bdtg]=[]
            tcs[bdtg].append([model,tau,flat,flon,fvmax,fe])


if(len(dtgs) == 0):
    print "WWWWWWWW no forecasts for: %s and storm: %s"%(tmodel,tstmid)
    ###sys.exit()

else:
    dtgs=mf.uniq(dtgs)

O=open(ftgpath,'w')

#
# check the which cycle JT is running in SHEM
#

if( (bid == 'P' or bid == 'S') and tmodel == 'ofc'):
    cnt00=0
    cnt06=0
    for dtg in dtgs:
        shour=dtg[8:]
        for t in tcs[dtg]:
            tau=t[1]
            if(tau == '000'):
                lat=float(t[2])
                if((shour == '00' or shour == '12') and lat > -80.0): cnt00=cnt00+1
                if((shour == '06' or shour == '18') and lat > -80.0): cnt06=cnt06+1
                
    print 'cccc 0000 ',cnt00
    print 'cccc 6666 ',cnt06

    if(cnt06 > cnt00): synopt='syn06_18'


if(synopt == 'syn00_12'):
    synhours=['00','12']
elif(synopt == 'syn06_18'):
    synhours=['06','18']

#
# set opt2 as synopt to g.tc.gs
#
opt2=synopt

for dtg in dtgs:
    csynhour=dtg[8:]
    
    doit=0
    for synhour in synhours:
        if(csynhour == synhour): doit=1

    if(not(doit)): continue

    card="%s %s ::"%(tmodel,dtg)
    n=0
    ltcs=len(tcs[dtg])
    for t in tcs[dtg]:
        (m,tau,lat,lon,vmax,fe)=t
        if(lat < 90.0 and int(tau) <= taumax and lat != -88.8):
            card="%s %s %5.1f %5.1f %s %5.0f"%(card,tau,lat,lon,vmax,fe)
            if(n != ltcs-1):
                card="%s :"%(card)
        
        n=n+1
        #print dtg,tau

    scard=card[0:len(card)-1]
    O.writelines(scard+'\n')

O.close()

siz=os.path.getsize(ftgpath)
print 'sssssssssssss ',siz
#
# if no tracks to plot,
#

plot='y'
vtype='hetero'
ptype='batch'
opt='null'
xsize='1024x768'
xsize='800x600'
vmodel=tmodel
nstorm=tstmid[0:2]
opt1="%s %s %s"%('name',tstmtype,tstmname)
nbasin=TC.Basin1toBasinNumber[bid]

gclinput="%s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s"%\
          (ptype,mdpath,ftgpath,vpath,pltdir,plot,vmodel,tstmid,nstorm,nbasin,veriname,vtype,bname,tyear,
           opt1,opt2,opt3)

gradsexe='grads'
if(prodopt):
    gradsopt='-lbc'
else:
    gradsopt='-lc'


looppath="%s/tc.trk.%s.%s.%s.%s.%s.loop.gif"%(pltdir,vmodel,tyear,tstmid,tstmname,tstmtype)
#tc.trk.ofc.2005.22P.INGRID.STY.png

cmd="rm /tmp/tc.fc.??.*"
mf.runcmd(cmd,ropt)

cmd="%s %s \"run g.tc.gs %s\" -g %s"%(gradsexe,gradsopt,gclinput,xsize)
mf.runcmd(cmd,ropt)

loopmask="/tmp/tc.fc.??.png"
pngs=glob.glob(loopmask)

delayloop=50
delaylast=200

n=1
npngs=len(pngs)
for png in pngs:

    if(n == 1):
        cmd="%s -loop 0 -delay %s /tmp/tc.fc.%02d.png "%(convertexe,delayloop,n)
    elif(n == npngs):
        cmd="%s -delay %s /tmp/tc.fc.%02d.png "%(cmd,delaylast,n)
    else:
        cmd="%s -delay %s /tmp/tc.fc.%02d.png "%(cmd,delayloop,n)
    n=n+1

cmd="%s %s"%(cmd,looppath)
mf.runcmd(cmd,ropt)


sys.exit()

File: prc/tc/g.tc.pl
#!/usr/bin/env perl

$perldir=$ENV{"WXMAP_PERL_DIR"};
require("$perldir/lib/mf.pl");

$wxdir=$ENV{"WXMAP_PRC_WXMAP_DIR"};
require("$wxdir/wxmap.env.pl");
wxmap_env();

$tcprcdir=$ENV{"WXMAP_PRC_TC_DIR"};

$tcprcdir="$tcprcdir/tctrack";
require("$tcprcdir/p.tc.func.pl");

tc_setup();
$plot='y';
$vtype='hetero';
$ptype='ops';
$opt1='NULL';
$mfilt='ASDFASDFASDF';
$bttype='ops.jtwc';

$verb=0;

$curdir=`pwd`;
chop($curdir);

#
# make 2-D stats from yearly .ctl file 
#

$narg=$#ARGV+1;

if($narg >= 5 ) {

  $i=0;
  $yyyy=$ARGV[$i]; $i++;
  $vvvv=$ARGV[$i]; $i++;
  $stmid=$ARGV[$i]; $i++;
  $vmodel=$ARGV[$i]; $i++;
  $ptype=$ARGV[$i]; $i++;
  $opt1=$ARGV[$i] if($ARGV[$i] ne '') ; $i++;
  if($i>=7) { $vtype=$ARGV[$i] ; $i++ }
  $mfilt=$ARGV[$i] if($ARGV[$i] ne '') ; $i++;

} else {

  print "$0:\n\n";
  print "Plots TC positions from the ERA and NCEP reanals with statistics\n";
  print "Uses the GrADS script g.tc.gs to do the plotting\n\n";
  print "\nThe argument list is:\n";
  print "        yyyy : YYYY | YYYYMM\n";
  print "       vvvv  : tmp | VVVV in  tc.veri.VVVV.txt file (e.g., drops.new)\n";
  print "       stmid : NNL where L = basin parameter\n";
  print "      vmodel : ifs | mmm1,mm2 -- verify model(s)\n";
  print "       ptype : [ops] | era40.batch | exp | rd | exp.ifs (current cases) | ops.batch\n";
  print "        opt1 : [NULL] | 2plot | ll.lat1.lat2.lon1.lon2\n";
  print "     [vtype] : [hetero] | homo\n";
  print "     [mfilt] : [] | models to filter out\n";
  print "\n";
  exit;
}


if(length($stmid) != 3) {
  print "EEE: invalid stmid: $stmid\n";
  exit(1);
}


$xsize='600x800';
$gmode="pbc";
$gmode="pc" if($plot eq "y");
$xsize="800x600";
$xsize="1024x768";
$gmode="lbc" if($ptype =~ 'batch');
$gmode="lc" if($plot eq "y" && !($ptype =~ 'batch') );
$gcmd='grads';


if($opt1 eq '2plot') {

  $xsize="600x800";
  $gmode="pbc";
  $gmode="pc" if($plot eq 'y');

}

if($ptype eq 'ops.batch' || $ptype eq 'era40.batch') {
  $gcmd='vgrads';
  $gcmd='grads';
}

if($opt1 =~ 'll') {
  @tt=split('\.',$opt1);
  $lat1=$tt[1];
  $lat2=$tt[2];
  $lon1=$tt[3]*1;
  $lon2=$tt[4]*1;
  $lon1=360+$lon1 if($lon1<0);
  $lon2=360+$lon2 if($lon2<0);

  $opt1="ll $lat1 $lat2 $lon1 $lon2";
}

#
# pass TC name to g.tc.gs
#
if($opt1 =~ 'name') {
  @tt=split('\.',$opt1);
  $tctype=$tt[1];
  $tcname=$tt[2];
  $opt1="name $tctype $tcname";
}

$bname=substr($stmid,2,1);
$bname=$basin{$bname};
$nbasin=$basin{substr($stmid,2,1),'num'};
$nstorm=substr($stmid,0,2)*1;

##################################################
#
# file setup
#
##################################################

$pltdir='/tmp';
$pltdir=$WXMAP{"WXMAP_TC_TRACK_PLOT_DIR"};


if($ptype =~ 'ops') {
  $btdir=$WXMAP{"WXMAP_TC_BT_OPS_DIR"};
  $ftdir=$WXMAP{"WXMAP_TC_FT_OPS_DIR"};
  $veridir="$WXMAP{'WXMAP_TC_VERI_OPS_DIR'}";
} elsif($ptype =~ 'exp' || $ptype =~ 'era40') {
  $btdir=$WXMAP{"WXMAP_TC_BT_EXP_DIR"};
  $ftdir=$WXMAP{"WXMAP_TC_FT_EXP_DIR"};
  $veridir="$WXMAP{'WXMAP_TC_VERI_EXP_DIR'}";


} else {
  $btdir=$WXMAP{"WXMAP_TC_BT_DIR"};
  $ftdir=$WXMAP{"WXMAP_TC_FT_DIR"};
  $veridir="$WXMAP{'WXMAP_TC_VERI_DIR'}";
}

if($vvvv =~ 'tmp') {
  $veridir="/tmp";
}

#ooooooooooooooooooooooooooooooooooooooooooooooooooo
#
# direct overide
#
#ooooooooooooooooooooooooooooooooooooooooooooooooooo

if($ptype eq 'era40' || $ptype eq 'era40.batch') {
  $btdir="/pcmdi/chico_dat/wxmap2/dat/tc/bt";
  $ftdir="/pcmdi/chico_dat/wxmap2/dat/tc/fc/era40";
  $veridir="/pcmdi/chico_dat/wxmap2/dat/tc/veri";
  $pltdir="/pcmdi/chico_dat/wxmap2/plt/tc/era40/trk";
}

if($ptype eq 'ops.batch') {
  $btdir="/pcmdi/chico_dat/wxmap2/dat/tc/bt";
  $ftdir="/pcmdi/tenki_dat/nwp/dat/tc/ft_ops";
  $veridir="/pcmdi/chico_dat/wxmap2/dat/tc/veri";
  $pltdir="/pcmdi/chico_dat/wxmap2/plt/tc/era40/trk";
}

print "PPPPPPPPPPPPPPP   btdir: $btdir\n";
print "PPPPPPPPPPPPPPP   ftdir: $ftdir\n";
print "PPPPPPPPPPPPPPP veridir: $veridir\n";

$grepyyyy=$yyyy;

if(length($yyyy) == 6) {

  $yyyymm=$yyyy;
  $yyyy=substr($yyyy,0,4);
  $grepyyyy=$yyyymm;

}

$btyyyy=2002;
$btyyyy=$yyyy;

if($btyyyy <= 2000 && !($ptype =~ 'era40')) {
    $bttype='neumann';
    $btdir="$btdir/bt_neumann";
}

if($bttype eq 'ops.jtwc') {
  $btpath="$btdir/$btyyyy/bt.jtwc.$bname.$btyyyy.$stmid.txt";
  $btpath="$btdir/$btyyyy/bt.local.jtwc.$bname.$yyyy.$stmid.txt";
} elsif($bttype eq 'neumann') {
  $btpath="$btdir/$btyyyy/bt.neumann.$bname.$btyyyy.$stmid.txt";
} else {
  $btpath="$btdir/$btyyyy/bt.ngtrp.$bname.$btyyyy.$stmid.txt";
}

if($ptype =~ 'exp' || $ptype =~ 'era40' ) {
  $ftipath="$veridir/tc.veri.$vvvv.ctat.txt";
  $vpath="$veridir/tc.veri.$vvvv.tc.ctl";
  $bname=substr($stmid,2,1);
  $bname=$basin_neumann{$bname};
  $btpath="$btdir/$yyyy/bt.neumann.$bname.$yyyy.$stmid.txt";
  if($yyyy > 2001) {
    $btdir=$WXMAP{"WXMAP_TC_BT_OPS_DIR"};
    $btpath="$btdir/$yyyy/bt.ngtrp.$bname.$yyyy.$stmid.txt";
  }

} else {
  $ftipath="$veridir/tc.veri.$vvvv.txt";
  $vpath="$veridir/tc.veri.$vvvv.tc.ctl";
}

#ooooooooooooooooooooooooooooooooooooooooooooooooooo
#
# direct overide
#
#ooooooooooooooooooooooooooooooooooooooooooooooooooo

if($ptype eq 'era40' || $ptype eq 'era40.batch' || $ptype eq 'ops.batch') {
    if($btyyyy >= 2001) {
	$bname=substr($stmid,2,1);
	$bname=$basin{$bname};
	$btpath="$btdir/$btyyyy/bt.local.jtwc.$bname.$btyyyy.$stmid.txt";
    } else {
	$btpath="$btdir/$btyyyy/bt.neumann.$bname.$btyyyy.$stmid.txt";
    }
    $ftipath="$veridir/tc.veri.$vvvv.txt";
}

if($ptype eq 'ops.w2') {

    $btdir="/pcmdi/chico_dat/wxmap2/dat/tc/bt";
    $ftdir="/pcmdi/chico_dat/wxmap2/dat/tc/fc/era40";
    $veridir="/pcmdi/chico_dat/wxmap2/dat/tc/veri";
    $pltdir="/pcmdi/chico_dat/wxmap2/plt/tc/era40/trk";
    $btpath="$btdir/$btyyyy/BT.$bname.$btyyyy.$stmid.txt";
    $ftipath="$veridir/tc.veri.$vvvv.txt";
}

$ftopath="/tmp/g.tc.ft.input.$yyyy.$stmid.txt";
$ftgpath='/tmp/g.tc.ft.txt';

print "nbtpath: $btpath\nftipath: $ftipath \nftopath: $ftopath\nftgpath: $ftgpath\n";

$overide=1;
$sizgft=(-s $ftopath);
if($sizgft eq '' || $sizgft == 0 || $overide) {
  $cmd="grep $grepyyyy $ftipath | grep $stmid | grep -v $mfilt > $ftopath";
  $cmd="grep $stmid $ftipath |  grep -v $mfilt > $ftopath";
  print "CCC($0): $cmd\n";
  system($cmd);
  $sizgft=(-s $ftopath);
#
# if SHEM look to previous year
#

  if($sizgft == 0 && ($bname eq 'sio' || $basin eq 'swp') ) {
    $grepyyyy=$grepyyyy-1;
    print "EEE: no forecast track cards to process\n";
    $cmd="grep $grepyyyy $ftipath | grep $stmid | grep -v $mfilt > $ftopath";
    $cmd="grep $stmid $ftipath |  grep -v $mfilt > $ftopath";
    print "CCC($0): $cmd\n";
    system($cmd);
    $sizgft=(-s $ftopath);
  }

}

$sizgft=(-s $ftopath);

if($sizgft == 0) {
  print "EEE: no forecast track cards to process\n";
  exit(1);
}

open(FI,$ftopath) || die "unable to open $ftopath";
open(FO,"> $ftgpath") || die "unable to open $ftgpath";


$dtgnew='00000000';
$modelnew='asdfasdf';
undef(@models);
@cards=<FI>;
foreach $card (@cards) {

  chomp($card);
  @tt=split(' ',$card);

  $model=$tt[0];
  $tau=$tt[1];
  $lat=$tt[6];
  $lon=$tt[7];
  $mw=$tt[16];
  $dtg=$tt[3];
  $hh=substr($dtg,8,2);

  $all{$model,$dtg,$tau,'lat'}=$lat;
  $all{$model,$dtg,$tau,'lon'}=$lon;
  $all{$model,$dtg,$tau,'mw'}=$mw;
  $all{$model,$dtg,'ntau'}++;

  push @models, $model;

# not really necessary ............... since the fc in the veri deck only has 12Z....
# plot only the 12Z forecast
#
#  print "ddddddddddddd $dtg\n";
#  if($ptype =~ 'era40' && $hh eq '12') {
#      push @dtgs, $dtg;
#  }  

  push @dtgs, $dtg;

  push @taus, $tau;

###  print "II: $model :: $dtg :: $tau :: $lat :: $lon :: $mw :: $hh\n";
  
}

@models=uniq(@models);
@dtgs=uniq(@dtgs);
@taus=uniq(@taus);

@tt=split(',',$vmodel);

if($#tt >= 1) {
  @models=@tt;
  $vmodel=$tt[0];
} 

if($#tt == 0) {
    undef(@models);
    push @models,$vmodel;
}


if($verb) {
    foreach $dtg (@dtgs) {
	print "DDD $dtg\n";
    }
}

foreach $model (@models) {
  foreach $dtg (@dtgs) {
    $fcard{$model,$dtg}='';

    foreach $tau (@taus) {
      $lat=$all{$model,$dtg,$tau,'lat'};
      $lon=$all{$model,$dtg,$tau,'lon'};
      $mw=$all{$model,$dtg,$tau,'mw'};
      if($tau eq '000' && $lon > 0) {
	$fcard{$model,$dtg}="$tau $lat $lon $mw";
      } elsif($lon > 0 && $lon < 360.0 && $tau gt '000' ) {
	$fcard{$model,$dtg}="$fcard{$model,$dtg} : $tau $lat $lon $mw "; 
      }
    }
  }
}

foreach $model (@models) {
  foreach $dtg (@dtgs) {
    if($fcard{$model,$dtg} ne '') {
      print "OO: $model $dtg :: $fcard{$model,$dtg}\n" if($verb);
      print FO "$model $dtg :: $fcard{$model,$dtg}\n";
    }
  }
}


close(FI);
close(FO);

$siz=(-s $ftgpath) if(-e $ftgpath) ;
print "dddddddddddddddddddddddd $ftopath\n";
print "DDDDDDDDDDDDDDDDDDDDDDDD $ftgpath :: $siz\n";

#sssssssssssssssssssssssssssssssssssssssssssss
#
#  sayoonara out if no fc tracks
#
#sssssssssssssssssssssssssssssssssssssssssssss

if($siz == 0) {
    print "EEEEEEEEEEEEEEEEEE no fc for: $vmodel\n";
    exit;
}

$gclinput="$ptype $btpath $ftgpath $vpath $pltdir $plot $vmodel $stmid $nstorm $nbasin $vvvv $vtype $bname $yyyy $opt1 $xsize";


$cmd="$gcmd -${gmode} \"run g.tc.gs $gclinput\" -g ${xsize}-0+0";
print "CCC(g.tc.gs) : $cmd\n";
#system($cmd);

exit;
 
system("rm $ftopath");
system("rm $ftgpath");
    
exit;

sub uniq (@list) {
  my(@mm,$m,$mnew);
  @list=@_;
  @mm=sort @list;
  undef(@list);
  $mnew="${mm[0]}ASDFASDFASDF";
  foreach $m (@mm) {
    if($m ne $mnew) {
      push @list,$m;
      $mnew=$m;
    }
  }
  return(@list);
}


